File: hoffman-agents-python/src/conscious_agent/core/experience_lexicon.py
from __future__ import annotations

from dataclasses import dataclass, field

import numpy as np


@dataclass
class LexiconEntry:
    label: str
    output_token: str
    trace_signature: np.ndarray
    prediction_error_peak: float = 0.0
    integration_depth: float = 0.5
    encounter_count: int = 0
    generation_bound: int = 0
    step_bound: int = 0
    labeling_source: str = "proto"
    associated_labels: list[str] = field(default_factory=list)


class ExperienceLexicon:
    def __init__(self, embedding_dim: int = 64, association_threshold: float = 0.15) -> None:
        if embedding_dim < 1:
            raise ValueError(f"embedding_dim must be >= 1, got {embedding_dim}")
        self._embedding_dim = embedding_dim
        self._association_threshold = association_threshold
        self._entries: dict[str, LexiconEntry] = {}
        self._signatures: list[tuple[np.ndarray, str]] = []

    def bind(
        self,
        label: str,
        trace_signature: np.ndarray,
        prediction_error_peak: float = 0.0,
        source: str = "proto",
        generation: int = 0,
        step: int = 0,
        output_token: str | None = None,
    ) -> LexiconEntry:
        trace_signature = self._normalize_signature(trace_signature)
        associations = self._compute_associations(trace_signature)

        if output_token is None:
            output_token = label

        entry = LexiconEntry(
            label=label,
            output_token=output_token,
            trace_signature=trace_signature,
            prediction_error_peak=prediction_error_peak,
            generation_bound=generation,
            step_bound=step,
            labeling_source=source,
            associated_labels=associations,
        )

        self._entries[label] = entry
        self._signatures.append((trace_signature, label))
        return entry

    def lookup_by_label(self, label: str) -> LexiconEntry | None:
        return self._entries.get(label)

    def lookup_by_signature(
        self,
        trace_signature: np.ndarray,
        threshold: float = 0.2,
    ) -> LexiconEntry | None:
        trace_signature = self._normalize_signature(trace_signature)

        best_dist = float("inf")
        best_label: str | None = None

        for sig, label in self._signatures:
            dist = float(np.linalg.norm(sig - trace_signature))
            if dist < best_dist:
                best_dist = dist
                best_label = label

        if best_label is not None and best_dist < threshold:
            return self._entries[best_label]
        return None

    def update_integration(self, label: str, encountered: bool) -> None:
        entry = self._entries.get(label)
        if entry is None:
            return

        delta_up = 0.05
        delta_down = 0.01

        if encountered:
            entry.integration_depth = min(entry.integration_depth + delta_up, 1.0)
            entry.encounter_count += 1
        else:
            entry.integration_depth = max(entry.integration_depth - delta_down, 0.0)

    def nearest_label(self, trace_signature: np.ndarray) -> tuple[str, float]:
        trace_signature = self._normalize_signature(trace_signature)

        if not self._signatures:
            return ("", 1.0)

        best_label = ""
        best_dist = float("inf")

        for sig, label in self._signatures:
            dist = float(np.linalg.norm(sig - trace_signature))
            if dist < best_dist:
                best_dist = dist
                best_label = label

        return (best_label, best_dist)

    def vocabulary_size(self) -> int:
        return sum(1 for e in self._entries.values() if e.integration_depth > 0.1)

    def sorted_by_integration(self) -> list[LexiconEntry]:
        return sorted(
            self._entries.values(),
            key=lambda e: e.integration_depth,
            reverse=True,
        )

    def _normalize_signature(self, sig: np.ndarray) -> np.ndarray:
        if sig.ndim == 0:
            arr = np.full(self._embedding_dim, float(sig))
        elif sig.ndim == 1 and sig.shape[0] < self._embedding_dim:
            arr = np.pad(sig, (0, self._embedding_dim - sig.shape[0]))
        elif sig.ndim == 1:
            arr = sig[:self._embedding_dim]
        else:
            arr = sig.flatten()[:self._embedding_dim]

        if arr.shape[0] < self._embedding_dim:
            arr = np.pad(arr, (0, self._embedding_dim - arr.shape[0]))

        norm = np.linalg.norm(arr)
        if norm > 0:
            arr = arr / norm
        return arr.astype(np.float64)

    def _compute_associations(self, trace_signature: np.ndarray) -> list[str]:
        trace_signature = self._normalize_signature(trace_signature)
        associations: list[tuple[float, str]] = []
        for sig, label in self._signatures:
            dist = float(np.linalg.norm(sig - trace_signature))
            if dist < self._association_threshold:
                associations.append((dist, label))

        associations.sort()
        return [label for _, label in associations[:10]]

    @property
    def embedding_dim(self) -> int:
        return self._embedding_dim

    @property
    def entry_count(self) -> int:
        return len(self._entries)

    def relabel(self, old_label: str, new_label: str, new_output_token: str | None = None) -> LexiconEntry | None:
        entry = self._entries.get(old_label)
        if entry is None:
            return None
        if new_output_token is None:
            new_output_token = new_label
        self._signatures = [(sig, lbl) for sig, lbl in self._signatures if lbl != old_label]
        del self._entries[old_label]
        entry.label = new_label
        entry.output_token = new_output_token
        entry.labeling_source = "parent"
        self._entries[new_label] = entry
        self._signatures.append((entry.trace_signature, new_label))
        return entry

    def remove(self, label: str) -> bool:
        if label not in self._entries:
            return False
        self._signatures = [(sig, lbl) for sig, lbl in self._signatures if lbl != label]
        del self._entries[label]
        return True

    def clear(self) -> None:
        self._entries.clear()
        self._signatures.clear()

    def to_dict(self) -> dict:
        entries = []
        for entry in self._entries.values():
            entries.append({
                "label": entry.label,
                "output_token": entry.output_token,
                "trace_signature": entry.trace_signature.tolist(),
                "prediction_error_peak": entry.prediction_error_peak,
                "integration_depth": entry.integration_depth,
                "encounter_count": entry.encounter_count,
                "generation_bound": entry.generation_bound,
                "step_bound": entry.step_bound,
                "labeling_source": entry.labeling_source,
                "associated_labels": list(entry.associated_labels),
            })
        return {
            "embedding_dim": self._embedding_dim,
            "association_threshold": self._association_threshold,
            "entries": entries,
        }

    @staticmethod
    def from_dict(data: dict) -> ExperienceLexicon:
        lex = ExperienceLexicon(
            embedding_dim=data.get("embedding_dim", 64),
            association_threshold=data.get("association_threshold", 0.15),
        )
        for ed in data.get("entries", []):
            sig = np.array(ed["trace_signature"], dtype=np.float64)
            entry = LexiconEntry(
                label=ed["label"],
                output_token=ed["output_token"],
                trace_signature=sig,
                prediction_error_peak=ed.get("prediction_error_peak", 0.0),
                integration_depth=ed.get("integration_depth", 0.5),
                encounter_count=ed.get("encounter_count", 0),
                generation_bound=ed.get("generation_bound", 0),
                step_bound=ed.get("step_bound", 0),
                labeling_source=ed.get("labeling_source", "proto"),
                associated_labels=list(ed.get("associated_labels", [])),
            )
            lex._entries[ed["label"]] = entry
            lex._signatures.append((sig, ed["label"]))
        return lex

File: hoffman-agents-react-native/src/core/experience-lexicon.js
class LexiconEntry {
  constructor(label, outputToken, traceSignature, {
    predictionErrorPeak = 0,
    integrationDepth = 0.5,
    encounterCount = 0,
    generationBound = 0,
    stepBound = 0,
    labelingSource = 'proto',
    associatedLabels = [],
  } = {}) {
    this.label = label;
    this.outputToken = outputToken;
    this.traceSignature = traceSignature;
    this.predictionErrorPeak = predictionErrorPeak;
    this.integrationDepth = integrationDepth;
    this.encounterCount = encounterCount;
    this.generationBound = generationBound;
    this.stepBound = stepBound;
    this.labelingSource = labelingSource;
    this.associatedLabels = associatedLabels;
  }
}

class ExperienceLexicon {
  constructor(embeddingDim = 64, associationThreshold = 0.15) {
    this._embeddingDim = embeddingDim;
    this._associationThreshold = associationThreshold;
    this._entries = new Map();
    this._signatures = [];
  }

  _normalizeSignature(sig) {
    let arr;
    if (typeof sig === 'number') {
      arr = new Float64Array(this._embeddingDim).fill(sig);
    } else if (sig.length < this._embeddingDim) {
      arr = new Float64Array(this._embeddingDim);
      arr.set(sig);
    } else {
      arr = sig.slice(0, this._embeddingDim);
    }

    let norm = 0;
    for (let i = 0; i < arr.length; i++) norm += arr[i] * arr[i];
    norm = Math.sqrt(norm);
    if (norm > 0) for (let i = 0; i < arr.length; i++) arr[i] /= norm;
    return arr;
  }

  _computeAssociations(traceSignature) {
    const dists = [];
    for (const [sig, label] of this._signatures) {
      let d = 0;
      for (let i = 0; i < sig.length; i++) d += (sig[i] - traceSignature[i]) ** 2;
      d = Math.sqrt(d);
      if (d < this._associationThreshold) dists.push([d, label]);
    }
    dists.sort((a, b) => a[0] - b[0]);
    return dists.slice(0, 10).map(([, l]) => l);
  }

  bind(label, traceSignature, {
    predictionErrorPeak = 0,
    source = 'proto',
    generation = 0,
    step = 0,
    outputToken = null,
  } = {}) {
    traceSignature = this._normalizeSignature(traceSignature);
    const associations = this._computeAssociations(traceSignature);
    if (!outputToken) outputToken = label;

    const entry = new LexiconEntry(label, outputToken, traceSignature, {
      predictionErrorPeak,
      generationBound: generation,
      stepBound: step,
      labelingSource: source,
      associatedLabels: associations,
    });

    this._entries.set(label, entry);
    this._signatures.push([traceSignature, label]);
    return entry;
  }

  lookupByLabel(label) { return this._entries.get(label) || null; }

  lookupBySignature(traceSignature, threshold = 0.2) {
    traceSignature = this._normalizeSignature(traceSignature);
    let bestDist = Infinity, bestLabel = null;
    for (const [sig, label] of this._signatures) {
      let d = 0;
      for (let i = 0; i < sig.length; i++) d += (sig[i] - traceSignature[i]) ** 2;
      d = Math.sqrt(d);
      if (d < bestDist) { bestDist = d; bestLabel = label; }
    }
    if (bestLabel && bestDist < threshold) return this._entries.get(bestLabel);
    return null;
  }

  updateIntegration(label, encountered) {
    const entry = this._entries.get(label);
    if (!entry) return;
    if (encountered) {
      entry.integrationDepth = Math.min(entry.integrationDepth + 0.05, 1);
      entry.encounterCount++;
    } else {
      entry.integrationDepth = Math.max(entry.integrationDepth - 0.01, 0);
    }
  }

  nearestLabel(traceSignature) {
    traceSignature = this._normalizeSignature(traceSignature);
    if (this._signatures.length === 0) return ['', 1];
    let bestLabel = '', bestDist = Infinity;
    for (const [sig, label] of this._signatures) {
      let d = 0;
      for (let i = 0; i < sig.length; i++) d += (sig[i] - traceSignature[i]) ** 2;
      d = Math.sqrt(d);
      if (d < bestDist) { bestDist = d; bestLabel = label; }
    }
    return [bestLabel, bestDist];
  }

  vocabularySize() {
    let count = 0;
    for (const entry of this._entries.values()) {
      if (entry.integrationDepth > 0.1) count++;
    }
    return count;
  }

  sortedByIntegration() {
    return [...this._entries.values()].sort((a, b) => b.integrationDepth - a.integrationDepth);
  }

  relabel(oldLabel, newLabel, newOutputToken) {
    const entry = this._entries.get(oldLabel);
    if (!entry) return null;
    if (!newOutputToken) newOutputToken = newLabel;
    this._signatures = this._signatures.filter(([, l]) => l !== oldLabel);
    this._entries.delete(oldLabel);
    entry.label = newLabel;
    entry.outputToken = newOutputToken;
    entry.labelingSource = 'parent';
    this._entries.set(newLabel, entry);
    this._signatures.push([entry.traceSignature, newLabel]);
    return entry;
  }

  remove(label) {
    if (!this._entries.has(label)) return false;
    this._signatures = this._signatures.filter(([, l]) => l !== label);
    this._entries.delete(label);
    return true;
  }

  clear() {
    this._entries.clear();
    this._signatures = [];
  }

  toDict() {
    const entries = [];
    for (const entry of this._entries.values()) {
      entries.push({
        label: entry.label,
        outputToken: entry.outputToken,
        traceSignature: Array.from(entry.traceSignature),
        predictionErrorPeak: entry.predictionErrorPeak,
        integrationDepth: entry.integrationDepth,
        encounterCount: entry.encounterCount,
        generationBound: entry.generationBound,
        stepBound: entry.stepBound,
        labelingSource: entry.labelingSource,
        associatedLabels: [...entry.associatedLabels],
      });
    }
    return { embeddingDim: this._embeddingDim, associationThreshold: this._associationThreshold, entries };
  }

  static fromDict(data) {
    const lex = new ExperienceLexicon(data.embeddingDim || 64, data.associationThreshold || 0.15);
    for (const ed of data.entries || []) {
      const sig = new Float64Array(ed.traceSignature);
      lex.bind(ed.label, sig, {
        predictionErrorPeak: ed.predictionErrorPeak || 0,
        source: ed.labelingSource || 'proto',
        generation: ed.generationBound || 0,
        step: ed.stepBound || 0,
        outputToken: ed.outputToken,
      });
      const entry = lex._entries.get(ed.label);
      if (entry) {
        entry.integrationDepth = ed.integrationDepth || 0.5;
        entry.encounterCount = ed.encounterCount || 0;
      }
    }
    return lex;
  }

  get embeddingDim() { return this._embeddingDim; }
  get entryCount() { return this._entries.size; }
}

module.exports = { ExperienceLexicon, LexiconEntry };
